src/components/ControlledAddressCombobox.tsx

```diff
@@ -0,0 +1,87 @@
+import { Combobox, Transition } from "@headlessui/react";
+import { Control, Controller, FieldError } from "react-hook-form";
+import { OnboardingFormInputs } from "../pages/profile";
+import { Fragment, SetStateAction } from "react";
+import { CarpoolFeature, CarpoolAddress } from "../utils/types";
+
+interface ControlledAddressComboboxProps {
+  control: Control<OnboardingFormInputs>;
+  name: "startAddress" | "companyAddress";
+  addressSelected: CarpoolAddress;
+  addressSetter: (val: SetStateAction<CarpoolAddress>) => void;
+  addressUpdater: (val: SetStateAction<string>) => void;
+  addressSuggestions: CarpoolFeature[];
+  error?: FieldError;
+}
+
+const ControlledAddressCombobox = (props: ControlledAddressComboboxProps) => {
+  return (
+    <Controller
+      name={props.name}
+      control={props.control}
+      render={({ field: { ref, ...fieldProps } }) => (
+        <Combobox
+          className={`w-full`}
+          as="div"
+          value={props.addressSelected}
+          onChange={(val: CarpoolFeature) => {
+            props.addressSetter(val);
+            fieldProps.onChange(val.place_name);
+          }}
+          ref={ref}
+        >
+          <Combobox.Input
+            className={`w-full shadow-sm rounded-md px-3 py-2 h-12 ${
+              props.error ? "border-northeastern-red" : "border-black"
+            }`}
+            displayValue={(feat: CarpoolAddress) =>
+              feat.place_name ? feat.place_name : ""
+            }
+            type="text"
+            onChange={(e) => {
+              if (e.target.value === "") {
+                props.addressSetter({
```

Comment:
Can you elaborate a little on the difference between the addressSetter and the addressUpdater? How are their purposes different?

Reply:
Not entirely sure, inherited code. I'll look into it later and see if they can be reduced into a single hook.

Reply:
Inherited code that i didn't think enough about. I'll see if they can be reduced to a single hook?

Reply:
the addressSetter and addressUpdater are useState setters used in Profile.tsx. This part of the code was inherited so I am not entirely sure. From my understanding, the addressUpdater uses a debounce, so that the API that gives us address suggestions is not spammed too much. In contrast, the addressSetter is what actually sets the form state to change to the address the user clicks on and selects.